src/ephemeris.c

```diff
@@ -0,0 +1,162 @@
+
+      /* Check if there is a new nav msg subframe to process.
+       * TODO: move this into a function */
+      if ((ch->state != TRACKING_RUNNING) ||
+          (ch->nav_msg.subframe_start_index == 0))
+        continue;
+
+      /* Decode ephemeris to temporary struct */
+      __asm__("CPSID i;");
+      s8 ret = process_subframe(&ch->nav_msg, &e);
+      __asm__("CPSIE i;");
+
+      if (ret <= 0)
+        continue;
+
+      /* Decoded a new ephemeris. */
+      if (memcmp(&es[ch->prn], &e, sizeof(e))) {
+        if ((e.iode != es[ch->prn].iode) || /* IODE has changed, as expected */
+            !es_confidence[ch->prn]) {      /* Or we lack confidnce... */
+          log_info("New ephemeris for PRN %02d\n", ch->prn+1);
+          /* Back up old ephemeris in case this turns out to be bad. */
+          chMtxLock(&es_mutex);
+          /* Use the new ephemeris without confidence. */
+          es_old[ch->prn] = es[ch->prn];
+          es[ch->prn] = e;
+          es_confidence[ch->prn] = false;
+          chMtxUnlock();
+        } else {
+          log_info("Ignoring new ephemeris for PRN %02d "
+                   "due to lack of confidence\n", ch->prn+1);
+        }
+      } else {
+        /* This is the second time we've decoded the same ephemeris, so
+         * we now have confidence. */
+        es_confidence[ch->prn] = true;
+      }
+
+      if (!es[ch->prn].healthy) {
+        log_info("PRN %02d unhealthy\n", ch->prn+1);
+      } else {
+        sbp_send_msg(SBP_MSG_EPHEMERIS, sizeof(ephemeris_t),
+                     (u8 *)&es[ch->prn]);
+      }
+    }
+  }
+
+  return 0;
+}
+
+static void ephemeris_msg_callback(u16 sender_id, u8 len, u8 msg[], void* context)
+{
+  (void)sender_id; (void)context;
+
+  if (len != sizeof(ephemeris_t)) {
+    log_warn("Received bad ephemeris from peer\n");
+    return;
+  }
+
+  ephemeris_t e = *(ephemeris_t *)msg;
+  if (e.prn >= MAX_SATS) {
+    log_warn("Ignoring ephemeris for invalid sat\n");
+    return;
+  }
+
+  if (memcmp(&e, &es[e.prn], sizeof(e)) == 0) {
+    /* Peer's ephemeris matches ours, we now have confidence. */
+    es_confidence[e.prn] = true;
+    return;
+  }
+
+  gps_time_t t = get_current_time();
+  if (!es_confidence[e.prn] && ephemeris_good(&e, t)) {
+    log_info("New ephemeris for PRN%d from peer\n", e.prn+1);
+    chMtxLock(&es_mutex);
+    es_old[e.prn] = es[e.prn];
+    es[e.prn] = e;
+    es_confidence[e.prn] = false;
+    chMtxUnlock();
+    
+  }
+}
+
+void ephemeris_setup(void)
+{
+  memset(es_confidence, 0, sizeof(es_confidence));
+  memset(es_old, 0, sizeof(es_old));
+  memset(es, 0, sizeof(es));
+  for (u8 i=0; i<32; i++) {
+    es[i].prn = i;
+  }
+
+  static sbp_msg_callbacks_node_t ephemeris_msg_node;
+  sbp_register_cbk(
+    SBP_MSG_EPHEMERIS,
+    &ephemeris_msg_callback,
+    &ephemeris_msg_node
+  );
+
+  chThdCreateStatic(wa_nav_msg_thread, sizeof(wa_nav_msg_thread),
+                    NORMALPRIO-1, nav_msg_thread, NULL);
```

Comment:
Check this thread is higher priority than `time_matched_obs` thread to help ensure it meets its timing deadline.